geopandas/tools/sjoin.py

```diff
@@ -186,6 +180,351 @@ def find_contains(a1, a2):
+    tree_idx = right_df.sindex
+
+    # validate max_search_neighbors and search_radius params
+    if max_search_neighbors is not None:
+        assert max_search_neighbors >= 1 and isinstance(
+            max_search_neighbors, int
+        ), "max_search_neighbors must be an integer and >= 1"
+        # warn about using rtree < 0.9.4 and max_search_neighbors option
+        # see https://github.com/Toblerity/rtree/pull/141
+
+        if [int(i) for i in RTREE_VERSION.split(".")] < [0, 9, 4]:
+            warn(
+                "Using an rtree version < 0.9.4 may cause inconsistent "
+                "results when using max_search_neighbors. Consider using a "
+                "large number or max_search_neighbors=None."
+            )
+    if search_radius is not None:
+        assert search_radius >= 0, "search_radius must be >=0"
+
+    def query_index(geo_in_l):
+        """
+        Queries the spatial index to filter results.
+        """
+        # restric by radius
+        if search_radius is not None:
+            if geo_in_l.geom_type == "Point" and search_radius == 0:
+                # applying .buffer(0) to a point returns an empty geometry
+                geo_search = geo_in_l.coords[0]
+            else:  # polygon or buffer > 0
+                geo_search = geo_in_l.buffer(search_radius).bounds
+            in_radius = list(tree_idx.intersection(geo_search))
+        else:
+            in_radius = right_df.index
+        # find neighbors
+        if max_search_neighbors is not None:
+            neighbors = list(
+                tree_idx.nearest(geo_in_l.bounds, num_results=max_search_neighbors)
+            )
+        else:
+            neighbors = right_df.index
+        # ensure indexes are int
+        return (int(idx) for idx in neighbors if idx in in_radius)
+
+    # the accuracy of the spatial index is limited, so we need to manually
+    # check each set of matches for actual distance
+    l_idx = []
+    r_idx = []
+    distances = []
+    for ind_in_left, geo_in_l in enumerate(left_df.geometry):
+        if not geo_in_l or geo_in_l.is_empty:
+            # see https://github.com/Toblerity/Shapely/issues/799
+            continue
+        possible_nearest = query_index(geo_in_l)
+        min_dist = np.inf  # initialize
+        min_ind = []  # initialize
+        for ind_in_right in possible_nearest:
+            geo_in_r = right_df.geometry.iloc[ind_in_right]
+            if not geo_in_r or geo_in_r.is_empty:
+                # see https://github.com/Toblerity/Shapely/issues/799
+                continue
+            dist = geo_in_l.distance(geo_in_r)
+            if dist < min_dist and (search_radius is None or dist <= search_radius):
+                # new closest
+                min_dist = dist  # reset
+                min_ind = [ind_in_right]  # re-initialize
+            elif dist == min_dist:
+                # matching closest, extend list of closest
+                min_ind.append(ind_in_right)
```

Comment:
I'd use `possible_nearest` to make a subset of gdf, then get all distances using vectorised `.distance` on GeoSeries and then `idxmin`. Especially .iloc done once will provide speedup, and with pygeos the distance as well.

Reply:
I tried something like this. Since we have to account for equidistant matches and such, I found that the performance actually dropped substantially. Maybe it makes sense if `possible_nearest` is expected to be large, but since the goal is for `possible_nearest` to be as small as possible in the first place, the overhead of going `lists -> numpy -> lists` outweighed any benefits.

I think that once pygeos is integrated with support for prepared geometries, both this and the original `sjoin` can be modified to get huge speedups.

Reply:
Not even .iloc done before and looping through the resulting gdf? I bet that would bring improvement here.

Reply:
Oh yeah, that is a free (no array creation) so why not. I'll implement that.

Reply:
I implemented applying `.iloc` before the loop, no issues. I think this is the best we can do until a faster vectorized `.distance` appears (I guess when `pygeos` get's merged)